### PR TITLE
HMSPROV-285: Reservation detail for AWS

### PR DESCRIPTION
**Pull request by lzap**
Although I thought for all filtering we would use URL parameters and provider type can be seen as a filter, because of OpenAPI, we must use the provider in the URL because payloads can be different only for different URLs in OpenAPI.

This adds a new endpoint `/reservations/TYPE/ID` that returns reservation with all the details loaded, currently it only works for AWS but the code is ready for extension.

I am leaving the old `/reservations/ID` still active, it returns only generic information and it can be useful for UI to get just basic info for the progress bar. Should be a tad faster as it does not need to join the two tables, so it's a better fit for polling.

There is still the open question for possible conflict between `TYPE` and `ID`, but I don't see this as a problem until someone complains about this, it's `GET` vs `POST` so it should be fine.

There are new `.http` files to test this feature.

### CODE REVIEW

File: internal/models/reservation_model.go [hunk 1]
```diff
@@ -74,10 +92,10 @@ type AWSReservation struct {
 	AWSReservationID string `db:"aws_reservation_id" json:"aws_reservation_id"`
 
 	// The ID of the image from which the instance is created. AMI's must be prefixed with 'ami-'.
-	ImageID string `json:"image_id"`
+	ImageID string `db:"image_id" json:"image_id"`
 
 	// Detail information is stored as JSON in DB
-	Detail *AWSDetail `db:"detail" json:"detail"`
+	Detail *AWSDetail `json:"detail"`
```
Comment (adiabramovitch) [inline internal/models/reservation_model.go:98]: Why did you remove the 'db'?

File: api/openapi.gen.json [hunk 2]
```diff
@@ -414,6 +414,40 @@
       }
     },
     "/reservations/aws": {
```
Comment (akhil-jha) [inline api/openapi.gen.json:416]: Missing {ID} I think.
`/reservations/aws/{ID}`

Comment (lzap) [reply api/openapi.gen.json:416]: Right, good catch. Fixing.

File: internal/models/reservation_model.go (hunk 1 above)
Comment (lzap) [reply internal/models/reservation_model.go:98]: Oh this was not intentional, fixing. I was playing around with this, it did not work well with pgx/sqlx and I left a TODO comment on two functions which I had to implement. If we used pgx/scanny natively it would work out of box as pgx driver supports all of the postgres native types (this is JSONB).

File: internal/dao/dao_interfaces.go [hunk 3]
```diff
@@ -73,6 +73,9 @@ type ReservationDao interface {
 	// GetById returns reservation for a particular account.
 	GetById(ctx context.Context, id int64) (*models.Reservation, error)
 
+	// GetAWSById returns reservation for a particular account.
+	GetAWSById(ctx context.Context, id int64) (*models.AWSReservation, error)
+
```
Comment (adiabramovitch) [inline internal/dao/dao_interfaces.go:78]: Maybe we should also add tickets for creating those methods for GCP and Azure?

File: internal/models/reservation_model.go [hunk 4]
```diff
@@ -61,6 +65,31 @@ type AWSDetail struct {
 	PowerOff bool `json:"poweroff"`
 }
 
+// TODO: Use pgx native driver with scany library instead of sqlx which does not
```
Comment (adiabramovitch) [inline internal/models/reservation_model.go:68]: Can you elaborate?

File: api/openapi.gen.json (hunk 2 above)
Comment (akhil-jha) [reply api/openapi.gen.json]: It did get fixed, and it's working,  but with the latest commit, I am getting the following for `GET /reservations/{ID}`
```
*** iqe_provisioning_api.exceptions.ApiTypeError: Invalid type for variable 'received_data'. Required value type is V1ReservationResponse and passed type was str at ['received_data']
```

Comment (lzap) [reply api/openapi.gen.json]: Ok this is a bug, working on this!

File: internal/dao/dao_interfaces.go (hunk 3 above)
Comment (lzap) [reply internal/dao/dao_interfaces.go:78]: Yeah, well, we will eventually figure this out when implementing these features :-)

File: internal/services/reservations_service.go [hunk 5]
```diff
@@ -55,31 +55,56 @@ func ListReservations(w http.ResponseWriter, r *http.Request) {
...
-func GetReservation(w http.ResponseWriter, r *http.Request) {
+func renderNotFoundOrDAOError(w http.ResponseWriter, r *http.Request, err error, daoError string) {
+	var e dao.NoRowsError
+	if errors.As(err, &e) {
+		renderError(w, r, payloads.NewNotFoundError(r.Context(), err))
+	} else {
+		renderError(w, r, payloads.NewDAOError(r.Context(), daoError, err))
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (lzap) [reply internal/services/reservations_service.go:67]: Absolutely, we already have `error_renderer.go` in that package, let's move it there.

**Comment by lzap** [comment]
Fixed the problem with the old endpoint, moved the code as @adiabramovitch suggested.

**Review by adiabramovitch — APPROVED** [review]
@lzap Thanks 😄 we need @RHEnVision/qe ack

**Comment by lzap** [comment]
@akhil-jha please merge once you confirm it works, thanks!

**Review by akhil-jha — APPROVED** [review]

File: internal/db/migrations/010_rename_provider_detail.sql [hunk 6]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE aws_reservation_details RENAME provider TO provider_fk;
```
Comment (ezr-ondrej) [inline internal/db/migrations/010_rename_provider_detail.sql:1]: Why are we renaming to `_fk`? Sounds like a weird col name, I'd expect constraint having this name

Comment (lzap) [reply internal/db/migrations/010_rename_provider_detail.sql:1]: There was a conflicts for the joined tables, otherwise we would need to drop "*" from the SQL replacing it with 13 columns. I thought I would rename instead of this.

Comment (ezr-ondrej) [reply internal/db/migrations/010_rename_provider_detail.sql:1]: ok, make sense, I'd still love less confusing name, but it's fine, not a big deal at all, just curious :) Thanks for the explanation :+1:
